Sources/MMIO/Documentation.docc/Essentials/Register-Blocks.md

```diff
@@ -0,0 +1,227 @@
+@RegisterBlock
+struct UARTPeripheral {
+    @RegisterBlock(offset: 0x00)
+    var data: Register<UARTData>
+
+    @RegisterBlock(offset: 0x04)
+    var status: Register<UARTStatus>
+
+    // The UARTControlBlock is located at offset 0x08
+    @RegisterBlock(offset: 0x08)
+    var controlBlock: UARTControlBlock
+}
+```
+
+The `@RegisterBlock(offset:)` macro works with both individual registers and other register blocks. When you access a register through a nested block, Swift MMIO automatically handles the address calculations:
+
+```swift
+// Create a UART instance at base address 0x40010000
+let uart = UARTPeripheral(unsafeAddress: 0x40010000)
+
+// Access the baud rate register through the nested block
+let baudRate = uart.controlBlock.baudRate
+```
+
+`uart.controlBlock.baudRate` points to address: `0x40010000 + 0x08 + 0x04 = 0x4001000C`
+
+### Working with repeated registers
+
+Many hardware peripherals contain repeated register structures. For example, a GPIO peripheral typically has identical configuration registers for each pin, or a timer might have multiple identical channels. Swift MMIO provides the ``MMIO/RegisterArray`` type to efficiently model these repeated structures.
+
+Let's enhance our UART example by adding support for multiple communication channels, where each channel has its own set of configuration registers:
+
+```swift
+import MMIO
+
+// Register definition for a single channel's configuration
+@Register(bitWidth: 32)
+struct ChannelConfig {
+    @ReadWrite(bits: 0..<1, as: Bool.self)
+    var enable: ENABLE
+
+    @ReadWrite(bits: 1..<3)
+    var priority: PRIORITY
+
+    @ReadWrite(bits: 3..<8)
+    var mode: MODE
+}
+```
+
+Now add an array of four channel configuration registers to the UART peripheral. Each register is 4 bytes apart from the next one, starting at offset 0x20:
+
+```swift
+@RegisterBlock
+struct UARTPeripheral {
+    @RegisterBlock(offset: 0x00)
+    var data: Register<UARTData>
+
+    @RegisterBlock(offset: 0x04)
+    var status: Register<UARTStatus>
+
+    @RegisterBlock(offset: 0x08)
+    var controlBlock: UARTControlBlock
+
+    @RegisterBlock(offset: 0x20, stride: 0x04, count: 4)
+    var channels: RegisterArray<ChannelConfig>
+}
+```
+
+The `channels` property has type ``MMIO/RegisterArray``, which is a type provided by Swift MMIO. Like Swift's standard arrays, you access elements using integer subscripts:
+
+```swift
+// Create a UART instance at base address 0x40010000
+let uart = UARTPeripheral(unsafeAddress: 0x40010000)
+
+// Access the configuration register for channel 2
+let channel2 = uart.channels[2]
+```
+
+When you access an array element, Swift MMIO automatically calculates its memory address:
+- `uart.channels[2]` points to address `0x40010000 + 0x20 + (2 * 0x04) = 0x40010028`
+
+> Important: Swift MMIO performs bounds checking on array accesses. Attempting to access `uart.channels[4]` would trigger a runtime trap since we defined the array with `count: 4` (valid indices are 0-3).
```

Comment:
```suggestion
> Important: Swift MMIO performs bounds checking on array accesses. Attempting to access `uart.channels[4]` would trigger a runtime trap since the example defined the array with `count: 4` (valid indices are 0-3).
```